components/board/pi/impl/digital_interrupts.go

```diff
@@ -0,0 +1,296 @@
+	return nil
+}
+
+func processFormula(oldFormula, newFormula, name string) (func(raw int64) int64, bool, error) {
+	if oldFormula == newFormula {
+		return nil, false, nil
+	}
+	x, err := scheme.Parse(newFormula)
+	if err != nil {
+		return nil, false, errors.Wrapf(err, "couldn't parse formula for %s", name)
+	}
+
+	testScope := scheme.Scope{}
+	num := 1.0
+	testScope["raw"] = &scheme.Value{Float: &num}
+	_, err = scheme.Eval(x, testScope)
+	if err != nil {
+		return nil, false, errors.Wrapf(err, "test exec failed for %s", name)
+	}
+
+	return func(raw int64) int64 {
+		scope := scheme.Scope{}
+		rr := float64(raw) // TODO(erh): fix
+		scope["raw"] = &scheme.Value{Float: &rr}
+		res, err := scheme.Eval(x, scope)
+		if err != nil {
+			panic(err)
+		}
+		f, err := res.ToFloat()
+		if err != nil {
+			panic(err)
+		}
+		return int64(f)
+	}, true, nil
+}
+
+// Reconfigure reconfigures this digital interrupt with a new formula.
+func (i *BasicDigitalInterrupt) Reconfigure(conf DigitalInterruptConfig) error {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	newFormula, isNew, err := processFormula(i.cfg.Formula, conf.Formula, conf.Name)
+	if err != nil {
+		return err
+	}
+	if !isNew {
+		return nil
+	}
+	i.pp = newFormula
+	i.cfg = conf
+	return nil
+}
+
+// A ServoDigitalInterrupt is an interrupt associated with a servo in order to
+// track the amount of time that has passed between low signals (pulse width). Post processors
+// make meaning of these widths.
+type ServoDigitalInterrupt struct {
+	last uint64
+	ra   *utils.RollingAverage
+
+	mu  sync.RWMutex
+	cfg DigitalInterruptConfig
+	pp  board.PostProcessor
+}
+
+// Value will return the window averaged value followed by its post processed
+// result.
+func (i *ServoDigitalInterrupt) Value(ctx context.Context, extra map[string]interface{}) (int64, error) {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+	v := int64(i.ra.Average())
+	if i.pp != nil {
+		return i.pp(v), nil
+	}
+
+	return v, nil
+}
+
+// Tick records the time between two successive low signals (pulse width). How it is
+// interpreted is based off the consumer of Value.
+func (i *ServoDigitalInterrupt) Tick(ctx context.Context, high bool, now uint64) error {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+	diff := now - i.last
+	i.last = now
+
+	if i.last == 0 {
+		return nil
+	}
+
+	if high {
+		// this is time between signals, ignore
+		return nil
+	}
+
+	i.ra.Add(int(diff / 1000))
+	return nil
+}
+
+// AddCallback currently panics.
+func (i *ServoDigitalInterrupt) AddCallback(c chan board.Tick) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	panic("servos can't have callback")
+}
+
+// RemoveCallback currently panics.
+func (i *ServoDigitalInterrupt) RemoveCallback(c chan board.Tick) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	panic("servos can't have callback")
+}
+
+// AddPostProcessor sets the post processor that will modify the value that
+// Value returns.
+func (i *ServoDigitalInterrupt) AddPostProcessor(pp board.PostProcessor) {
```

Comment:
Do these need to be exported functions at all anymore?

Reply:
Hm... I'm leaning towards "yes," but am not confident of that. This one is public because that's part of the `board.DigitalInterrupt` interface. It's used publicly by the fake board implementation, and it wouldn't surprise me if users of the Go SDK relied on that part of the interface, too. but maybe it could be made private and removed from the interface, and it's just a coincidence that 2 of the 3 implementations of the interface have these?

I'm inclined to just merge this as-is and think that through more another time, but I'm open to thinking it through now if you prefer.

Reply:
Add a todo - unexport AddPostProcessor function from `pi/impl/digital_intterupts.go` and a jira ticket for further clean up if you still prefer doing that.

Reply:
I looked into putting it back, and realized it would be easier to just roll forward and remove it entirely. I still wish I'd broken this up into 2 or perhaps 3 smaller PRs.